lib/open_project/authentication.rb

```diff
@@ -61,14 +73,68 @@ def values
       end
     end
 
-    module Realm
+    ##
+    # Options used in the WWW-Authenticate header returned to the user
+    # in case authentication failed (401).
+    module WWWAuthenticate
       module_function
 
-      def realm
+      def pick_auth_scheme(supported_schemes, default_scheme, request_headers = {})
+        key = request_headers.keys.find { |h| h =~ /X-Authentication-Scheme$/i }
```

Comment:
Do we really have to parse request headers ourselves? If there is no other way, `.starts_with?` is [probably faster](http://stackoverflow.com/questions/4168543/how-do-i-find-if-a-string-starts-with-an-other-string-in-ruby).